File: src/external/Bass/c/tempo/tempo.c
/*****************************************************************************
 tempo.c/h/rc - Copyright (c) 2003-2008 (: JOBnik! :) [Arthur Aminov, ISRAEL]
                                                      [http://www.jobnik.org]
                                                      [   bass_fx@jobnik.org]
 
 BASS_FX tempo / rate / pitch with dsp fx
 * Imports: bass.lib, bass_fx.lib
            kernel32.lib, user32.lib, comdlg32.lib, gdi32.lib
*****************************************************************************/

#include <windows.h>
#include <commctrl.h>
#include <stdio.h>
#include "bass.h"
#include "bass_fx.h"
#include "tempo.h"

HWND win=NULL;
HINSTANCE inst;

HSTREAM chan;			// tempo channel handle
HFONT Font;				// font handle
HFX fxEQ;				// dsp peaking eq handle
float oldfreq;			// old sample rate

#define MESS(id,m,w,l) SendDlgItemMessage(win,id,m,(WPARAM)w,(LPARAM)l)

OPENFILENAME ofn;
char path[MAX_PATH];

// display error dialogs
void Error(const char *es)
{
	char mes[200];
	sprintf(mes,"%s\n\n(error code: %d)",es,BASS_ErrorGetCode());
	MessageBox(win,mes,"Error",MB_ICONEXCLAMATION);
}

// update dsp eq
void UpdateFX(int b)
{
	BASS_BFX_PEAKEQ eq;
	int v = MESS(IDC_SLDEQ1+b,TBM_GETPOS,0,0);
    
	eq.lBand = b;	// get values of the selected band
    BASS_FXGetParameters(fxEQ, &eq);
		eq.fGain = (float)(10 - v);
	BASS_FXSetParameters(fxEQ, &eq);
}

// set dsp eq
void SetDSP_EQ(float fGain, float fBandwidth, float fQ, float fCenter_Bass, float fCenter_Mid, float fCenter_Treble)
{
	BASS_BFX_PEAKEQ eq;

	// set peaking equalizer effect with no bands
	fxEQ=BASS_ChannelSetFX(chan, BASS_FX_BFX_PEAKEQ,0);

	eq.fGain=fGain;
	eq.fQ=fQ;
	eq.fBandwidth=fBandwidth;
	eq.lChannel=BASS_BFX_CHANALL;

	// create 1st band for bass
	eq.lBand=0;
	eq.fCenter=fCenter_Bass;
	BASS_FXSetParameters(fxEQ, &eq);

	// create 2nd band for mid
	eq.lBand=1;
	eq.fCenter=fCenter_Mid;
	BASS_FXSetParameters(fxEQ, &eq);

	// create 3rd band for treble
	eq.lBand=2;
	eq.fCenter=fCenter_Treble;
	BASS_FXSetParameters(fxEQ, &eq);

	// update dsp eq
	UpdateFX(0);
	UpdateFX(1);
	UpdateFX(2);
}

// show the approximate position in MM:SS format
void UpdatePositionLabel(void)
{
	if(!BASS_FX_TempoGetRateRatio(chan)) return;
	{
		char c[30];
		float totalsec=(float)MESS(IDC_POS,TBM_GETRANGEMAX,0,0)/BASS_FX_TempoGetRateRatio(chan);
		float posec=(float)MESS(IDC_POS,TBM_GETPOS,0,0)/BASS_FX_TempoGetRateRatio(chan);
		sprintf(c,"Playing position: %02d:%02d / %02d:%02d", (int)posec/60,(int)posec%60,(int)totalsec/60,(int)totalsec%60);
		MESS(IDC_SPOS,WM_SETTEXT,0,c);
	}
}

// get the file name from the file path
char *GetFileName(const char *filepath)
{
	unsigned char slash_location;
	filepath = strrev(filepath);
	slash_location = strchr(filepath, '\\') - filepath;
	return (strrev(filepath) + strlen(filepath) - slash_location);
}

BOOL CALLBACK dialogproc(HWND h,UINT m,WPARAM w,LPARAM l)
{
	DWORD p;
	float freq;
	char c[30];

	switch (m) {
		case WM_COMMAND:
			switch (LOWORD(w)) {
				case ID_OPEN:
					{
						char file[MAX_PATH]="";
						ofn.lpstrFilter="playable files\0*.mo3;*.xm;*.mod;*.s3m;*.it;*.mtm;*.mp3;*.mp2;*.mp1;*.ogg;*.wav;*aif\0All files\0*.*\0\0";
						ofn.lpstrFile=file;
						if (GetOpenFileName(&ofn)) {
							memcpy(path,file,ofn.nFileOffset);
							path[ofn.nFileOffset-1]=0;

							// free previous streams/dsp handles
							BASS_StreamFree(chan);

							// create decode stream channel
							chan=BASS_StreamCreateFile(FALSE, file, 0, 0, BASS_STREAM_DECODE);

							// create decode music channel
							if (!chan) chan=BASS_MusicLoad(FALSE, file, 0, 0, BASS_MUSIC_RAMP|BASS_MUSIC_PRESCAN|BASS_STREAM_DECODE,0);

							if (!chan){
								// not a WAV/MP3 or MOD
								MESS(ID_OPEN,WM_SETTEXT,0,"click here to open a file && play it...");
								Error("Selected file couldn't be loaded!");
								break;
							}

							// update the position slider
							p=(DWORD)BASS_ChannelBytes2Seconds(chan, BASS_ChannelGetLength(chan, BASS_POS_BYTE));
							MESS(IDC_POS,TBM_SETRANGEMAX,0,p);
							MESS(IDC_POS,TBM_SETPOS,TRUE,0);

							// get the current sample rate
							BASS_ChannelGetAttribute(chan, BASS_ATTRIB_FREQ, &freq);
							oldfreq = freq;

							// create a new stream - decoded & resampled :)
							if (!(chan=BASS_FX_TempoCreate(chan, BASS_SAMPLE_LOOP|BASS_FX_FREESOURCE))){
								MESS(ID_OPEN,WM_SETTEXT,0,"click here to open a file && play it...");
								Error("Couldn't create a resampled stream!");
								BASS_StreamFree(chan);
								BASS_MusicFree(chan);
								break;
							}

							// set dsp eq to channel
							SetDSP_EQ(0.0f, 2.5f, 0.0f, 125.0f, 1000.0f, 8000.0f);

							// update the Button to show the loaded file name
							MESS(ID_OPEN,WM_SETTEXT,0,GetFileName(file));

							// set Volume
							p=MESS(IDC_VOL,TBM_GETPOS,0,0);
							BASS_ChannelSetAttribute(chan, BASS_ATTRIB_VOL, (float)p/100.0f);

							// update tempo sliders
							MESS(IDC_RATE,TBM_SETRANGEMAX,0,(long)(freq * 1.3f));
							MESS(IDC_RATE,TBM_SETRANGEMIN,0,(long)(freq * 0.7f));
							MESS(IDC_RATE,TBM_SETPOS,TRUE,(long)freq);
							MESS(IDC_RATE,TBM_SETPAGESIZE,0,(long)(freq * 0.01f));		// by 1%

							sprintf(c,"Samplerate = %dHz", (long)freq);
							MESS(IDC_SRATE,WM_SETTEXT,0,c);

							MESS(IDC_TEMPO,TBM_SETPOS,TRUE,0);
							MESS(IDC_STEMPO,WM_SETTEXT,0,"Tempo = 0%");

							MESS(IDC_PITCH,TBM_SETPOS,TRUE,0);
							MESS(IDC_SPITCH,WM_SETTEXT,0,"Pitch Scaling = 0 semitones");

							// update the approximate time in seconds view
							UpdatePositionLabel();

							// play new created stream
							BASS_ChannelPlay(chan,FALSE);
						}
					}
					return 1;
			}
			break;

		case WM_VSCROLL:
			{
				if(l){
					UpdateFX(GetDlgCtrlID((HWND)l)-IDC_SLDEQ1);
				}
			}
			break;

		case WM_HSCROLL:
			{
				if(!BASS_ChannelIsActive(chan)) break;

				switch (GetDlgCtrlID((HWND)l)) {
					case IDC_TEMPO:
							// set new tempo
							BASS_ChannelSetAttribute(chan, BASS_ATTRIB_TEMPO, (float)MESS(IDC_TEMPO, TBM_GETPOS, 0, 0));

							// update tempo static text
							sprintf(c,"Tempo = %d%%", MESS(IDC_TEMPO, TBM_GETPOS, 0, 0));
							MESS(IDC_STEMPO,WM_SETTEXT,0,c);
					case IDC_RATE:
							// set new samplerate
							BASS_ChannelSetAttribute(chan, BASS_ATTRIB_TEMPO_FREQ, (float)MESS(IDC_RATE, TBM_GETPOS, 0, 0));

							// update samplerate static text
							sprintf(c,"Samplerate = %dHz", MESS(IDC_RATE, TBM_GETPOS, 0, 0));
							MESS(IDC_SRATE,WM_SETTEXT,0,c);

							// update all bands fCenters after changing samplerate
							{
								BASS_BFX_PEAKEQ eq;
								int i;

								for(i=0;i<3;i++){
									eq.lBand = i;
									BASS_FXGetParameters(fxEQ, &eq);
										eq.fCenter = eq.fCenter * (float)MESS(IDC_RATE, TBM_GETPOS, 0, 0) / oldfreq;
									BASS_FXSetParameters(fxEQ, &eq);
								}
								oldfreq = (float)MESS(IDC_RATE, TBM_GETPOS, 0, 0);
							}
					case IDC_PITCH:
						{
							// set new pitch scale
							BASS_ChannelSetAttribute(chan, BASS_ATTRIB_TEMPO_PITCH, (float)MESS(IDC_PITCH, TBM_GETPOS, 0, 0));

							// update pitch static text
							sprintf(c,"Pitch Scaling = %d semitones", MESS(IDC_PITCH, TBM_GETPOS, 0, 0));
							MESS(IDC_SPITCH,WM_SETTEXT,0,c);

							// update the approximate time in seconds view
							UpdatePositionLabel();
						}
					break;
					case IDC_POS:
						// change the position
						if (LOWORD(w) == SB_ENDSCROLL) { // seek to new pos
							BASS_ChannelSetPosition(chan, (QWORD)BASS_ChannelSeconds2Bytes(chan, (double)MESS(IDC_POS,TBM_GETPOS,0,0)), BASS_POS_BYTE);
						}
						// update the approximate time in seconds view
						UpdatePositionLabel();
					break;
					case IDC_VOL:
						BASS_ChannelSetAttribute(chan, BASS_ATTRIB_VOL, (float)MESS(IDC_VOL,TBM_GETPOS,0,0)/100.0f);
					break;
				}
			}
			return 1;
			break;

		case WM_CLOSE:
			KillTimer(h,1);
			EndDialog(h, 0);
			return 1;
		break;

		case WM_INITDIALOG:
			win=h;
			GetCurrentDirectory(MAX_PATH,path);
			memset(&ofn,0,sizeof(ofn));
			ofn.lStructSize=sizeof(ofn);
			ofn.hwndOwner=h;
			ofn.hInstance=inst;
			ofn.nMaxFile=MAX_PATH;
			ofn.Flags=OFN_HIDEREADONLY|OFN_EXPLORER;

			// setup output - default device, 44100hz, stereo, 16 bits
			if (!BASS_Init(-1,44100,0,win,NULL)) {
				Error("Can't initialize device");
				DestroyWindow(win);
				return 1;
			}

			// initialize dsp eq sliders
			MESS(IDC_SLDEQ1,TBM_SETRANGE,FALSE,MAKELONG(0,20));
			MESS(IDC_SLDEQ1,TBM_SETPOS,TRUE,10);

			MESS(IDC_SLDEQ2,TBM_SETRANGE,FALSE,MAKELONG(0,20));
			MESS(IDC_SLDEQ2,TBM_SETPOS,TRUE,10);

			MESS(IDC_SLDEQ3,TBM_SETRANGE,FALSE,MAKELONG(0,20));
			MESS(IDC_SLDEQ3,TBM_SETPOS,TRUE,10);

			// volume range
			MESS(IDC_VOL,TBM_SETRANGEMAX,0,100);
			MESS(IDC_VOL,TBM_SETPOS,TRUE,50);
			MESS(IDC_VOL,TBM_SETPAGESIZE,0,5);

			// tempo
			MESS(IDC_TEMPO,TBM_SETRANGEMAX,TRUE,30);
			MESS(IDC_TEMPO,TBM_SETRANGEMIN,TRUE,-30);
			MESS(IDC_TEMPO,TBM_SETPOS,TRUE,0);
			MESS(IDC_TEMPO,TBM_SETPAGESIZE,0,1);

			// rate
			MESS(IDC_RATE,TBM_SETRANGEMAX,0,(long)(44100.0f * 1.3f));
			MESS(IDC_RATE,TBM_SETRANGEMIN,0,(long)(44100.0f * 0.7f));
			MESS(IDC_RATE,TBM_SETPOS,TRUE,44100);
			MESS(IDC_RATE,TBM_SETPAGESIZE,0,(long)(44100.0f * 0.01f));	// by 1%

			// pitch
			MESS(IDC_PITCH,TBM_SETRANGEMAX,TRUE,30);
			MESS(IDC_PITCH,TBM_SETRANGEMIN,TRUE,-30);
			MESS(IDC_PITCH,TBM_SETPOS,TRUE,0);
			MESS(IDC_PITCH,TBM_SETPAGESIZE,0,1);

			// set bold font for IDC_SDSPEQ
			Font=CreateFont(-11,0,0,0,FW_BOLD,FALSE,FALSE,FALSE,ANSI_CHARSET,OUT_STROKE_PRECIS,
							CLIP_STROKE_PRECIS,DRAFT_QUALITY,DEFAULT_PITCH|FF_DONTCARE,"MS Sans Serif");
			MESS(IDC_SDSPEQ,WM_SETFONT,Font,0);
			return 1;
	}
	return 0;
}

int PASCAL WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,LPSTR lpCmdLine, int nCmdShow)
{
	inst=hInstance;

	// check the correct BASS was loaded
	if (HIWORD(BASS_GetVersion())!=BASSVERSION) {
		MessageBox(0,"An incorrect version of BASS.DLL was loaded (2.4 is required)","Incorrect BASS.DLL",MB_ICONERROR);
		return 1;
	}

	// check the correct BASS_FX was loaded
	if (HIWORD(BASS_FX_GetVersion())!=BASSVERSION) {
		MessageBox(0,"An incorrect version of BASS_FX.DLL was loaded (2.4 is required)","Incorrect BASS_FX.DLL",MB_ICONERROR);
		return 1;
	}

	DialogBox(inst,(char*)1000,0,&dialogproc);

	BASS_Free();

	DeleteObject(Font);

	return 0;
}

File: src/external/Bass/delphi/Tempo/tempo.pas
{==========================================================================
 tempo.pas - Copyright (c) 2003-2008 (: JOBnik! :) [Arthur Aminov, ISRAEL]
                                                   [http://www.jobnik.org]
                                                   [   bass_fx@jobnik.org]

 BASS_FX tempo / pitch / rate with dsp fx
 * Requires: BASS 2.4 (available @ www.un4seen.com)
==========================================================================}

unit tempo;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ComCtrls, BASS, BASS_FX;

type
  TfrmTempo = class(TForm)
    GroupBox: TGroupBox;
    lbl125Hz: TLabel;
    lbl1KHz: TLabel;
    lbl8KHz: TLabel;
    lblTempo: TLabel;
    lblSamplerate: TLabel;
    lblPitch: TLabel;
    lblPosition: TLabel;
    vsEQBASS: TScrollBar;
    vsEQMID: TScrollBar;
    vsEQTREBLE: TScrollBar;
    tbTempo: TTrackBar;
    tbPitch: TTrackBar;
    tbSamplerate: TTrackBar;
    btnOpen: TButton;
    od: TOpenDialog;
    tbVolume: TTrackBar;
    lblVolume: TLabel;
    lblDSPEQ: TLabel;
    hsPosition: TScrollBar;
    procedure FormCreate(Sender: TObject);
    procedure tbVolumeChange(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure btnOpenClick(Sender: TObject);
    procedure lblTempoClick(Sender: TObject);
    procedure lblSamplerateClick(Sender: TObject);
    procedure lblPitchClick(Sender: TObject);
    procedure tbTempoChange(Sender: TObject);
    procedure tbSamplerateChange(Sender: TObject);
    procedure tbPitchChange(Sender: TObject);
    procedure hsPositionChange(Sender: TObject);
    procedure hsPositionScroll(Sender: TObject; ScrollCode: TScrollCode; var ScrollPos: Integer);
    procedure vsEQBASSChange(Sender: TObject);
    procedure vsEQMIDChange(Sender: TObject);
    procedure vsEQTREBLEChange(Sender: TObject);
    procedure Error(msg: string);
    procedure UpdatePositionLabel();
    procedure UpdateFX(band, pos: integer);
    procedure SetDSP_EQ(fGain, fBandwidth, fQ, fCenter_Bass, fCenter_Mid, fCenter_Treble: FLOAT);
  private
    { Private declarations }
    chan: DWORD;               // tempo channel handle
    fxEQ: DWORD;               // dsp peaking eq handle
    eq: BASS_BFX_PEAKEQ;       // dsp peaking eq
    info: BASS_CHANNELINFO;    // channel info
    freq: FLOAT;               // sample rate
    oldfreq: FLOAT;            // old sample rate
  public
    { Public declarations }
  end;

var
  frmTempo: TfrmTempo;

implementation

{$R *.DFM}

// display error dialogs
procedure TFrmTempo.Error(msg: string);
var s: string;
begin
    s := msg + #13#10 + #13#10 + '(error code: ' + IntToStr(BASS_ErrorGetCode) + ')';

  MessageBox(handle, PChar(s), 'Error', MB_ICONEXCLAMATION);
end;

procedure TfrmTempo.FormCreate(Sender: TObject);
begin
  // check the correct BASS was loaded
  If (HiWord(BASS_GetVersion()) <> BASSVERSION) then begin
    application.MessageBox('An incorrect version of BASS.DLL was loaded (2.4 is required)', 'Incorrect BASS.DLL', MB_ICONERROR or MB_OK);
    application.Terminate;
  end;

  // check the correct BASS_FX was loaded
  If (HiWord(BASS_FX_GetVersion()) <> BASSVERSION) then begin
    application.MessageBox('An incorrect version of BASS_FX.DLL was loaded (2.4 is required)', 'Incorrect BASS_FX.DLL', MB_ICONERROR or MB_OK);
    application.Terminate;
  end;

  // setup output - default device, 44100hz, stereo, 16 bits
  If BASS_Init(-1, 44100, 0, frmTempo.handle, nil) = false then begin
    Error('Couldnt Initialize Digital Output');
    application.Terminate;
  end;

  freq := 44100;
end;

procedure TfrmTempo.FormDestroy(Sender: TObject);
begin
  BASS_Free;    // free BASS & BASS_FX
end;

procedure TfrmTempo.tbVolumeChange(Sender: TObject);
begin
  BASS_ChannelSetAttribute(chan, BASS_ATTRIB_VOL, tbVolume.position / 100);
end;

procedure TfrmTempo.btnOpenClick(Sender: TObject);
begin
  od.Filename := '';
  od.Filter :=  'Playable files|*.mo3;*.xm;*.mod;*.s3m;*.it;*.mtm;*.mp3;*.mp2;*.mp1;*.ogg;*.wav;*.aif|All files|*.*';
  If od.Execute = false then exit;

  // free previous streams/music/dsps
  BASS_StreamFree(chan);

  // create decode channel
  chan := BASS_StreamCreateFile(FALSE, pChar(od.filename), 0, 0, BASS_STREAM_DECODE);

  // check for MOD
  If chan = 0 Then chan := BASS_MusicLoad(FALSE, pChar(od.filename), 0, 0, BASS_MUSIC_RAMP or BASS_MUSIC_PRESCAN or BASS_MUSIC_DECODE, 0);

  If chan = 0 then begin
    btnOpen.Caption := 'click here to open a file && play it...';
    Error('Selected file couldnt be loaded!');
    Exit;
  end;

  // get channel info
  BASS_ChannelGetInfo(chan, info);

  freq := info.freq;
  oldfreq := freq;

  // set MAX length to hsPosition Scroller in seconds
  hsPosition.max := round(BASS_ChannelBytes2Seconds(chan, BASS_ChannelGetLength(chan, BASS_POS_BYTE)));
  hsPosition.position := 0;

  // create a new stream - decoded & resampled :)
  chan := BASS_FX_TempoCreate(chan, BASS_SAMPLE_LOOP or BASS_FX_FREESOURCE);
  If (chan = 0) Then begin
    btnOpen.Caption := 'click here to open a file && play it...';
    Error('Couldn''t create a resampled stream!');
    BASS_StreamFree(chan);
    BASS_MusicFree(chan);
    Exit;
  End;

  // set dsp eq to channel
  SetDSP_EQ(0, 2.5, 0, 125, 1000, 8000);

  // update the Button to show the loaded file name
  btnOpen.Caption := ExtractFileName(pChar(od.Filename));

  // set Volume
  tbVolume.OnChange(Self);

  // set rate min/max values according to current frequency
  With tbSamplerate do begin
    min := 0;    // to avoit min > max
    max := 1;    // to avoid max < min
    max := round(freq * 1.3);
    min := round(freq * 0.7);
    pagesize := round(freq * 0.01);      // by 1%
  End;

  // update the Rate label hint
  lblSamplerate.hint := 'click to set Samplerate to ' + FloatToStr(freq) + 'Hz';

  // update the approximate time in seconds view
  UpdatePositionLabel;

  // play new created stream
  BASS_ChannelPlay(chan, FALSE);

  // set tempo/rate/pitch to normal
  lblTempo.OnClick(Self);
  lblSamplerate.OnClick(Self);
  lblPitch.OnClick(Self);
end;

procedure TfrmTempo.lblTempoClick(Sender: TObject);
begin
  tbTempo.position := 0;
  tbTempo.OnChange(Self);
end;

procedure TfrmTempo.lblSamplerateClick(Sender: TObject);
begin
  tbSamplerate.position := Round(freq);
  tbSamplerate.OnChange(Self);
end;

procedure TfrmTempo.lblPitchClick(Sender: TObject);
begin
  tbPitch.position := 0;
  tbPitch.OnChange(Self);
end;

procedure TfrmTempo.tbTempoChange(Sender: TObject);
begin
  If (BASS_ChannelIsActive(chan) = 0) Then Exit;

  BASS_ChannelSetAttribute(chan, BASS_ATTRIB_TEMPO, tbTempo.position);
  lblTempo.Caption := 'Tempo = ' + IntToStr(tbTempo.position) + '%';

  // update the approximate time in seconds view
  UpdatePositionLabel;
end;

procedure TfrmTempo.tbSamplerateChange(Sender: TObject);
var i: DWORD;
begin
  If (BASS_ChannelIsActive(chan) = 0) Then Exit;

  BASS_ChannelSetAttribute(chan, BASS_ATTRIB_TEMPO_FREQ, tbSamplerate.position);
  lblSamplerate.Caption := 'Samplerate = ' + IntToStr(tbSamplerate.position) + 'Hz';

  // update all bands fCenters after changing samplerate
  for i := 0 to 2 do begin
    eq.lBand := i;
    BASS_FXGetParameters(fxEQ, @eq);
      eq.fCenter := eq.fCenter * tbSamplerate.position / oldfreq;
    BASS_FXSetParameters(fxEQ, @eq);
  end;

  oldfreq := tbSamplerate.position;

  // update the approximate time in seconds view
  UpdatePositionLabel;
end;

procedure TfrmTempo.tbPitchChange(Sender: TObject);
begin
  If (BASS_ChannelIsActive(chan) = 0) Then Exit;

  BASS_ChannelSetAttribute(chan, BASS_ATTRIB_TEMPO_PITCH, tbPitch.position);
  lblPitch.Caption := 'Pitch Scaling = ' + IntToStr(tbPitch.position) + 'semitones';

  // update the approximate time in seconds view
  UpdatePositionLabel;
end;

procedure TfrmTempo.hsPositionChange(Sender: TObject);
begin
  If (BASS_ChannelIsActive(chan) = 0) then exit;
  // update the approximate time in seconds view
  UpdatePositionLabel;
end;

procedure TfrmTempo.hsPositionScroll(Sender: TObject; ScrollCode: TScrollCode; var ScrollPos: Integer);
begin
  If (BASS_ChannelIsActive(chan) = 0) then exit;
  if ScrollCode = scEndScroll then BASS_ChannelSetPosition(chan, BASS_ChannelSeconds2Bytes(chan, hsPosition.position), BASS_POS_BYTE);
  // update the approximate time in seconds view
  UpdatePositionLabel;
end;

procedure TfrmTempo.vsEQBASSChange(Sender: TObject);
begin
  UpdateFX(0, vsEQBASS.Position);
end;

procedure TfrmTempo.vsEQMIDChange(Sender: TObject);
begin
  UpdateFX(1, vsEQMID.Position);
end;

procedure TfrmTempo.vsEQTREBLEChange(Sender: TObject);
begin
  UpdateFX(2, vsEQTREBLE.Position);
end;

// Set DSP Peaking EQ
procedure TfrmTempo.SetDSP_EQ(fGain, fBandwidth, fQ, fCenter_Bass, fCenter_Mid, fCenter_Treble: FLOAT);
begin
  // set peaking equalizer effect with no bands
  fxEQ := BASS_ChannelSetFX(chan, BASS_FX_BFX_PEAKEQ, 0);

  eq.fGain := fGain;
  eq.fQ := fQ;
  eq.fBandwidth := fBandwidth;
  eq.lChannel := BASS_BFX_CHANALL;

  // create 1st band for bass
  eq.lBand := 0;
  eq.fCenter := fCenter_Bass;
  BASS_FXSetParameters(fxEQ, @eq);

  // create 2nd band for mid
  eq.lBand := 1;
  eq.fCenter := fCenter_Mid;
  BASS_FXSetParameters(fxEQ, @eq);

  // create 3rd band for treble
  eq.lBand := 2;
  eq.fCenter := fCenter_Treble;
  BASS_FXSetParameters(fxEQ, @eq);

  // update dsp eq
  UpdateFX(0, vsEQBASS.position);
  UpdateFX(1, vsEQMID.position);
  UpdateFX(2, vsEQTREBLE.position);
End;

// update dsp peaking eq
procedure TfrmTempo.UpdateFX(band, pos: integer);
begin
  eq.lBand := band;     // get values of the selected band
  BASS_FXGetParameters(fxEQ, @eq);
    eq.fGain := 10 - pos;
  BASS_FXSetParameters(fxEQ, @eq);
End;

// show the approximate position in MM:SS format
procedure TfrmTempo.UpdatePositionLabel();
var totalsec, posec: FLOAT;
begin
  if(BASS_FX_TempoGetRateRatio(chan)>0) then begin
    totalsec := hsPosition.max / BASS_FX_TempoGetRateRatio(chan);
    posec := hsPosition.position / BASS_FX_TempoGetRateRatio(chan);

    lblPosition.Caption := 'Playing position: ' +
          format('%2.2d:%2.2d', [trunc(posec) div 60, trunc(posec) mod 60]) + ' / ' +
          format('%2.2d:%2.2d', [trunc(totalsec) div 60, trunc(totalsec) mod 60]);
  end;
end;

end.
